src/frequenz/sdk/microgrid/client.py

```diff
@@ -297,7 +315,8 @@ async def _component_data_task(
             )
             try:
                 call = self.api.GetComponentData(
-                    microgrid_pb.ComponentIdParam(id=component_id)
+                    microgrid_pb.ComponentIdParam(id=component_id),
+                    timeout=timeout,
```

Comment:
I'm wondering if this timeout for the data stream will 
  - kill the stream after the timeout (this is undesirable, but we might not notice because we have retry code to restart the connection)
  - or kill the stream only if it hasn't transmitted any data for a while.

Have you tried this out already?

Reply:
I haven't tried it yet. Will give it a go.

Reply:
FYI @leandro-lucarella-frequenz @ela-kotulska-frequenz, we discussed this with @sahas-subramanian-frequenz and decided to remove the timeout from the methods for getting component data, because the `GetComponentData` gets killed whenever a timeout error occurs and this block is re-executed:
https://github.com/frequenz-floss/frequenz-sdk-python/blob/v0.x.x/src/frequenz/sdk/microgrid/client.py#L299-L303

and this might cause some performance issues.

The method gets killed, because grpc timeout error falls into `grpc.aio.AioRpcError `category. If we don’t want to retry on timeout errors, we could inspect the status code of that `grpc.aio.AioRpcError` and take an appropriate action, if its status is equal to `grpc.StatusCode.DEADLINE_EXCEEDED`.

The consumer keeps consuming messages like so `async for msg in receiver` without any interruption or errors.

I checked that using the test function below (it has been removed from the code, because there's no timeout applied to `GetComponentData` anymore, but leaving it for reference). What it does is:
- it mocks the `GrpcServer` to keep sending some random battery data
- sending the 5th message on the server is delayed on purpose to cause a timeout error on the client side
- the client keeps consuming messages from the stream until it consumed some messages after the timeout already occurred

```python
@patch("frequenz.sdk.microgrid.client.DEFAULT_GRPC_CALL_TIMEOUT", GRPC_CALL_TIMEOUT)
async def test_get_component_data_stream(mocker: MockerFixture) -> None:
    """Test if the component data stream survives timeout errors"""
    servicer = MockMicrogridServicer()
    servicer.add_component(1, ComponentCategory.COMPONENT_CATEGORY_BATTERY)

    delayed_msg_index = 3

    # pylint: disable=unused-argument
    async def mock_component_data(
        request: ComponentIdParam, context: Any
    ) -> AsyncGenerator[ComponentData, None]:
        counter = 0
        while True:
            if counter == delayed_msg_index:
                await asyncio.sleep(GRPC_SERVER_DELAY)
            else:
                await asyncio.sleep(GRPC_CALL_TIMEOUT / 3)
            counter += 1
            timestamp = Timestamp()
            timestamp.GetCurrentTime()
            yield ComponentData(
                id=request.id,
                ts=timestamp,
                battery=Battery(
                    data=BatteryData(
                        soc=MetricAggregation(avg=random.uniform(1.0, 100.0)),
                    )
                ),
            )

    mocker.patch.object(servicer, "GetComponentData", mock_component_data)
    server = MockGrpcServer(servicer, port=57809)
    await server.start()

    target = "[::]:57809"
    grpc_channel = grpc.aio.insecure_channel(target)
    client = MicrogridGrpcClient(grpc_channel=grpc_channel, target=target)

    receiver = await client.battery_data(component_id=1)
    messages: List[int] = []
    while len(messages) < delayed_msg_index + 2:
        msg = await receiver.receive()
        messages.append(msg.component_id)

    await server.wait_for_termination(GRPC_SERVER_SHUTDOWN_DELAY)
```

Reply:
So basically with streams there is no way to tell if a connection stopped getting data because there is really no data to transmit from the producing end or if the produced is actually stalled right? But if, for example, the connection is closed, we'll get and error and restart the streaming, correct?

If so, it sounds completely reasonable. I guess the only way around this is to enable some keep-alive mechanism, but probably not doable using gRPC as I would guess streams are one way only. We could have a `ping` call to see if the whole API server is up and responding on time, but beyond that it is probably too overkill.

Reply:
Yeah, good point. I think you are right that the streams are expected to produce data relatively frequently and if it doesn't happen for a longer time it perceived as a timeout on the client side, whether there was no data or if the device was broken.

Regarding the connection being closed:
- if a timeout error occurs, then the retry logic is executed and the stream is restored
- but the connection might also be closed by the producer? I remember there was an option to send a `None` to indicate the end of the stream? I think in this case the stream wouldn't be recreated

Yeah, the keep-alive mechanism sounds like something that could be useful in the future.

Reply:
The other concern was that if we keep reestablishing our streaming calls,  we'll end up missing datapoints that were sent between the timeout and the retry interval.  That would be highly undesirable,  even though the reconnection will be transparent to the user.